File: couchdb-external-hook.rb
#
# This is the external process hook script that forwards requests from CouchDB
# to couchdb-lucene.  It requires the following:
#
# * A Ruby interpreter (http://www.ruby-lang.org/en/downloads/)
# * The RubyGems package management system (http://docs.rubygems.org/read/chapter/3)
# * The 'json' ruby gem (sudo gem install json)
#
# == Configuration options
# couchdb.lucene.host=mymachine.somehost.com - The name of the machine hosting
#    the couchdb-lucene server (default is localhost)
# couchdb.lucene.port=1234 - The port the couchdb-lucene is running on
#    (default is 5985)
# couchdb.log.dir=/some/log/dir - The directory you wish to use to store the
#    logs from this script (default is /var/log/couchdb)
#
# == Usage
# Include the following in your CouchDB local.ini file in the [external] section:
# fti=/path/to/ruby /usr/lib/couchdb/couchdb-lucene/couchdb-external-hook.rb [options]
#
# Example:
# fti=/path/to/ruby /usr/lib/couchdb/couchdb-lucene/couchdb-external-hook.rb couchdb.log.dir=/some/log/dir couchdb.lucene.port=1234
#
# == Author
# John Wood
# john_p_wood@yahoo.com
# http://johnpwood.net
#

require 'net/http'
require 'zlib'
require 'rubygems'
require 'json'


DEFAULT_LOG_DIR = '/var/log/couchdb/'
DEFAULT_LUCENE_HOST = 'localhost'
DEFAULT_LUCENE_PORT = 5985
LOGFILE = 'couchdb-external-hook.log'


def args
  @@arg_array ||= begin
    args = {}
    ARGV.each do |token|
      key, value = token.strip.split('=')
      args[key] = value
    end
    args
  end
end

def log(message)
  t = Time.now
  @@log << t.strftime("%Y-%m-%d %H:%M:%S") + ":#{t.usec}" << " :: " << message << "\n"
  @@log.flush
end

def logfile
  @@logfile ||= begin
    logdir = args['couchdb.log.dir'] || DEFAULT_LOG_DIR
    logdir << '/' unless logdir =~ /\/$/
    logdir + LOGFILE
  end
end

def respond_with_error(code, message)
  puts ({"code" => code, "body" => message}).to_json
  STDOUT.flush
  true
end

def parse_path(path)
  [path[0], path[2], path[3]]
end

def build_query_string(input_hash)
  couch_query = input_hash['query']
  if couch_query.nil?
    respond_with_error(400, "No query found in request.") and return
  end

  database, design_doc, view = parse_path(input_hash['path'])
  if database.nil? ||design_doc.nil? || view.nil?
    respond_with_error(400, 'Path must contain datbase name, design doc name, and view name.') and return
  end

  command = couch_query['q'].nil? ? 'info' : 'search' 
  lucene_query = "/#{command}/#{database}/#{design_doc}/#{view}?"
  couch_query.each do |name, value|
    lucene_query << "#{name}=#{value}&"
  end
  lucene_query.sub!(/[&?]$/, '')    

  log "couchdb-lucene query: #{lucene_query}"
  lucene_query
end

def send_query_to_couchdb_lucene(input_hash, query_string)
  @@host ||=  args['couchdb.lucene.host'] || DEFAULT_LUCENE_HOST
  @@port ||= (args['couchdb.lucene.port'] || DEFAULT_LUCENE_PORT).to_i

  connection = Net::HTTP.new(@@host, @@port)
  resp, data = connection.get(query_string, input_hash['headers'])

  data = inflate(data)
  resp.delete("Content-Encoding")

  log "couchdb-lucene response: (#{resp.code}) #{data}"

  if resp.code.to_i == 200
    headers = {}
    resp.each_capitalized {|k,v| headers[k] = v}

    response = {"code" => resp.code, "headers" => headers}
    resp.content_type =~ /json/ ? response['json'] = data : response['body'] = data
    
    log "Response to CouchDB: #{response.to_json}"
    puts response.to_json
    STDOUT.flush
  else
    respond_with_error(resp.code, resp.message)
  end
rescue Exception => e
  log "Error encountered when sending request to couchdb-lucene: #{e.inspect}"
  respond_with_error(500, "Unexpected error when sending request to couchdb-lucene.")
end

def inflate(string)
  gz = Zlib::GzipReader.new(StringIO.new(string))
  gz.read
rescue
  string
end

def execute_query(line)
  log "Request from CouchDB: #{line}"
  input_hash = JSON.parse(line)
  query_string = build_query_string(input_hash)
  send_query_to_couchdb_lucene(input_hash, query_string) unless query_string.nil?
end


#
# Entry point
#
File.open(logfile, 'a') do |log|
  @@log = log
  log "Searcher started."
  
  execute_query(STDIN.gets)  # Uncomment for testing; reloads script every time

#  while (line = STDIN.gets)   #
#    execute_query(line)       # Comment out for testing; loads script once
#  end                         #
end

File: couchdb-external-hook.py
#!/usr/bin/python

import httplib
import optparse as op
import sys
import traceback
import urllib

try:
    import json
except:
    import simplejson as json

__usage__ = "%prog [OPTIONS]"

httpdict = {"etag": "ETag", "content-type": "Content-Type"}


def options():
    return [
        op.make_option('--remote-host', dest='remote_host',
                       default="localhost",
                       help="Hostname of the couchdb-lucene server. [%default]"),
        op.make_option('--remote-port', dest='remote_port', type='int',
                       default=5985,
                       help="Port of the couchdb-lucene server. [%default]"),
        op.make_option('--local-key', dest='key',
                       default="local",
                       help="Configured key name for this couchdb instance. [%default]"),
    ]


def main():
    parser = op.OptionParser(usage=__usage__, option_list=options())
    opts, args = parser.parse_args()

    if len(args):
        parser.error("Unrecognized arguments: %s" % ' '.join(args))
    for req in requests():
        res = httplib.HTTPConnection(opts.remote_host, opts.remote_port)
        try:
            resp = respond(res, req, opts.key)
        except Exception, e:
            body = traceback.format_exc()
            resp = mkresp(500, body, {"Content-Type": "text/plain"})

        res.close()
        sys.stdout.write(json.dumps(resp))
        sys.stdout.write("\n")
        sys.stdout.flush()


def requests():
    line = sys.stdin.readline()
    while line:
        yield json.loads(line)
        line = sys.stdin.readline()


def respond(res, req, key):
    path = req.get("path", [])

    # Drop name of external hook.
    del path[1]

    # URL-escape each part
    for index, item in enumerate(path):
        path[index] = urllib.quote(path[index], safe="")

    path = '/'.join(['', key] + path)
    params = urllib.urlencode(
        dict([k, v.encode('utf-8')] for k, v in req["query"].items()))
    path = '?'.join([path, params])

    req_headers = {}
    for h in req.get("headers", []):
        if h.lower() in ["accept", "if-none-match"]:
            req_headers[h] = req["headers"][h]

    # verb renamed to method in 0.11 onwards.
    if "method" in req:
        method = req["method"]
    else:
        method = req["verb"]

    res.request(method, path, headers=req_headers)
    resp = res.getresponse()

    resp_headers = {}
    for h, v in resp.getheaders():
        if h.lower() in httpdict:
            resp_headers[httpdict[h]] = resp.getheader(h, [])

    return mkresp(resp.status, resp.read(), resp_headers)


def mkresp(code, body, headers=None):
    ret = {"code": code, "body": body}
    if headers is not None:
        ret["headers"] = headers
    return ret


if __name__ == "__main__":
    main()

